src/SmartComponents/Drift/utils.js

```diff
@@ -0,0 +1,76 @@
+export function formatDate(date) {
+    const d = new Date(date);
+    return new Date(d.getTime() - d.getTimezoneOffset() * 60 * 1000).toISOString().split('T')[0];
+}
+
+;
+
+export function getDate(ago) {
+    let d = new Date();
+    d.setDate(d.getDate() - ago);
+    return formatDate(d);
+}
+
+export function parseDriftData(data) {
+    const payload = {};
+    payload.numEvents = data.meta.count;
+    payload.baselineDetail = [];
+    let unique_baselines = [];
+
+    data.data.forEach((item) => groupPayload(item.payload, payload.baselineDetail, unique_baselines));
```

Comment:
This is really complicated, can you give us a bit of information what is the input and what should be on the output? I guess it can be solved with a simple `reduce` function or a `map` but I don't fully understand the structure of the `data`.

Reply:
The input data is as follow:

```
{
data: [
{
id: "77667272-e6c1-4ab7-b32a-6d6f554b7fc4",
created: "2022-01-13T07:49:14.792547",
bundle: "Red Hat Enterprise Linux",
application: "Drift",
event_type: "Drift from baseline detected",
payload: "{"bundle": "rhel", "application": "drift", "event_type": "drift-baseline-detected", "timestamp": "2022-01-13T07:49:13.982999+00:00", "account_id": "6089719", "events": [{"metadata": {}, "payload": "{\"baseline_id\": \"313c0813-552e-481f-9128-d6c687ddd3fb\", \"baseline_name\": \"baseline-msager-tags\"}"}], "context": "{\"inventory_id\": \"04fb3dc1-a43c-4b24-994d-09e7ee44fa69\", \"system_check_in\": \"2022-01-13T07:49:10.982091+00:00\", \"display_name\": \"msager-rhel84\", \"tags\": [{\"key\": \"group\", \"value\": \"msager-test-clients\", \"namespace\": \"insights-client\"}]}"}",
actions: [
{
id: "f6ed32eb-9084-42b0-b207-05adbdbfd1ba",
endpoint_type: "webhook",
invocation_result: false,
endpoint_id: "847d6e49-6142-4283-83ce-828db3f9663b",
details: null
},
{
id: "5ed335fd-85b3-428a-b97a-b8d486827635",
endpoint_type: "email_subscription",
invocation_result: true,
endpoint_id: "49b180c3-17f2-49af-af74-33ea8495b66b",
details: null
},
{
id: "dae1d5e5-323a-40c6-a983-97de89799292",
endpoint_type: "email_subscription",
invocation_result: true,
endpoint_id: "49b180c3-17f2-49af-af74-33ea8495b66b",
details: null
}
]
},
{
id: "64c0446f-87a1-4af9-b4df-b32850f18e72",
created: "2022-01-13T06:36:37.992885",
bundle: "Red Hat Enterprise Linux",
application: "Drift",
event_type: "Drift from baseline detected",
payload: "{"bundle": "rhel", "application": "drift", "event_type": "drift-baseline-detected", "timestamp": "2022-01-13T06:36:37.149843+00:00", "account_id": "6089719", "events": [{"metadata": {}, "payload": "{\"baseline_id\": \"313c0813-552e-481f-9128-d6c687ddd3fb\", \"baseline_name\": \"baseline-msager-tags\"}"}], "context": "{\"inventory_id\": \"ba950140-17e3-426d-aea2-498b9718fb4d\", \"system_check_in\": \"2022-01-13T06:36:36.945861+00:00\", \"display_name\": \"msager-rhel84-obf\", \"tags\": [{\"key\": \"group\", \"namespace\": \"insights-client\", \"value\": \"msager-test-clients\"}]}"}",
actions: [
{
id: "079ba3e3-3d9c-401b-b187-3821ca4aa276",
endpoint_type: "email_subscription",
invocation_result: true,
endpoint_id: "49b180c3-17f2-49af-af74-33ea8495b66b",
details: null
},
...
]
}
],
links: {
last: "/api/notifications/v1.0/notifications/events?limit=20&offset=0",
first: "/api/notifications/v1.0/notifications/events?limit=20&offset=0"
},
meta: {
count: 15
}
}
```

I need to get what is in the `payload` key and group this data to achieve something like this:
```
{
    "numEvents": 15,
    "baselineDetail": [
        {
            "baselineId": "313c0813-552e-481f-9128-d6c687ddd3fb",
            "baselineName": "baseline-msager-tags",
            "systems": [
                "04fb3dc1-a43c-4b24-994d-09e7ee44fa69",
                "ba950140-17e3-426d-aea2-498b9718fb4d",
                "bd60cb31-46fe-46fb-b580-7ced02fbf5d0"
            ]
        },
        {
            "baselineId": "9a6568db-b670-4cd9-9e44-284bccaad0e1",
            "baselineName": "automation_baseline_311_24d55299-73b0-4c1a-bed2-c8d5e1b95c9f",
            "systems": [
                "54b89b3f-017c-4643-91da-0ce19dd53a95",
                "e9d35787-9929-4a66-8b63-6f7775437844",
                "58dcfd91-4e57-4c4c-b8ff-2ed9cd277d3e"
            ]
        },
        {
            "baselineId": "f859a765-25f4-4b51-946d-23e7bfbf69d8",
            "baselineName": "automation_baseline_311_0bea93d4-ecac-40c6-a25e-b6b45ddf4d9d",
            "systems": [
                "e9d35787-9929-4a66-8b63-6f7775437844",
                "54b89b3f-017c-4643-91da-0ce19dd53a95"
            ]
        },
        {
            "baselineId": "864c532e-7792-41e2-9bdd-4acf593e9cc2",
            "baselineName": "baseline-msager-test-clients",
            "systems": [
                "bd60cb31-46fe-46fb-b580-7ced02fbf5d0"
            ]
        },
        {
            "baselineId": "f569b1ee-efc4-4550-9e17-72d72e50ccd3",
            "baselineName": "automation_baseline_311_4222b9e1-d553-4d86-9f89-060520b3dadb",
            "systems": [
                "54b89b3f-017c-4643-91da-0ce19dd53a95"
            ]
        },
        {
            "baselineId": "5745ada6-40f4-4903-b7f5-2192d4a6107a",
            "baselineName": "automation_baseline_311_250ed013-df68-4ea8-93c3-bfdf7bfffd69",
            "systems": [
                "54b89b3f-017c-4643-91da-0ce19dd53a95"
            ]
        }
    ]
}
```

That's is the purpose of `parseDriftData` method

Reply:
Thank you for the input and output. As I suspected this can be simplified with use of `reduce`. I'd like to avoid nested `for`s, but that's probably not possible :disappointed: 
```JS
const parseDriftData = (data) => ({
  numEvents: data.meta.count,
  baselineDetail: groupPayload(data.data)
  .sort(({systems: aSystems}, { systems: bSystems }) => aSystems.length < bSystems.length ?
    1 :
    (aSystems.length > bSystems.length ? -1 : 0)
  ),
})

const groupPayload = (data) => {
  return data.reduce((acc, curr) => {
    // We'll use just events and inventoryId from context
    const { events, context: { inventory_id: inventoryId } } = parsePayload(curr.payload);
​
    // Let's iterate over all events (and use just the event)
    events.forEach(({ payload: event }) => {
      // Is there already baselineId?
      const { systems } = acc.find(({ baselineId }) => baselineId === event.baseline_id) || {};
​
      // either assign to list of systems, or create new record
      if (systems) {
         !systems.includes(inventoryId) && systems.push(inventoryId)
      } else {
         acc.push({
          baselineId: event.baseline_id,
          baselineName: event.baseline_name,
          systems: [inventoryId]
         });
      }
    });
​
    return acc;
  }, [])
}
```

You can further improve it by adding a sort function based on systems, or adjusting the reduce function. The best thing about this is that `parseDriftData` function has not variables and further `groupPayload` doesn't have any side effects (it's not mutating argument `array` sent from `parseDriftData`). You could change the `forEach` to `map` and you could return the value directly from `reduce`, which could improve the speed of this function. But there's a drawback in readability and you'd probably have to inject previous values which would introduce yet another branching.